### PR TITLE
merge for PR

### PR DESCRIPTION
**Pull request by Alon6**

### CODE REVIEW

File: BST/BST_task.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+from flask import Flask, request
+from pymongo import MongoClient
+from Tree import Tree
+
+
+# Initializing server and BST variables
+app = Flask(__name__)
+client = MongoClient("localhost", 27017)
+db = client["bst_db"]
+table = db["bst_table12"]
```
Comment (ofekba) [inline BST/BST_task.py]: what if the dbq tabe not exists?

File: BST/BST_task.py [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+bst = Tree()
+
+# Function empties MongoDB and the in-program tree
+@app.route('/delete_all_treasures', methods=['DELETE'])
+def delete_all_treasures():
+    table.delete_many({})
```
Comment (ofekba) [inline BST/BST_task.py]: what will happened if the delete action will fail?

File: BST/BST_task.py [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+    bst.root = None
+    return "Success", 200
+
+@app.route('/insert_treasure', methods=['POST'])
+def insert_treasure():
+    val_json = request.get_json()
+    val = val_json["value"]
```
Comment (ofekba) [inline BST/BST_task.py]: if there if not key "value" in the input Json what could happend?

File: BST/BST_task.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+    if bst.insert(Node(val), table):
+        return "Success", 200
```
Comment (ofekba) [inline BST/BST_task.py]: return HttpStatus.OK constant instead of hard coded code + text

File: BST/BST_task.py [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+    else:
+        return "Error", 400
```
Comment (ofekba) [inline BST/BST_task.py]: what is the error? why returns 400? 
add error handling

File: BST/BST_task.py [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+def search_treasure():
+    val = request.args.get("value")
+    if bst.search(val):
+        return {"message" : "Treasure found!"}, 200
+    else:
+        return {"message" : "Treasure not found"}, 400
+
+@app.route('/pre_order_traversal', methods=['GET'])
+def pre_order_traversal():
+    val = bst.bst_pass("pre-order")
```
Comment (ofekba) [inline BST/BST_task.py]: why not using constants or separate to different functions?

File: BST/BST_task.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+    if val is None:
+        return "Error", 400
+    else:
+        return {"traversal_result": val}, 200
+
+@app.route('/in_order_traversal', methods=['GET'])
+def in_order_traversal():
+    val = bst.bst_pass("in-order")
+    if val is None:
```
Comment (ofekba) [inline BST/BST_task.py]: use shorten writing

File: BST/BST_task.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+    else:
+        return {"message" : "BST is not valid"}, 400
+
+if __name__ == "__main__":
+    app.run()
+
+
+"""
+bst = rebuild(table)
+    get_vis(table)
```
Comment (ofekba) [inline BST/BST_task.py:145]: write test with pytest + test cases
unit tests + e2e
cover how much edge cases + valid input checking as you can.

File: BST/Node.py [hunk 9]
```diff
@@ -0,0 +1,84 @@
+
+import math
+from math import inf
+class Node:
+    # Initializing empty node
+    def __init__(self, value):
+        self.left = None
+        self.right = None
+        self.treasure = value
+    # Function returns the path to a node if exist else return the path to the place in which the node will be added
```
Comment (ofekba) [inline BST/Node.py]: the doc for each function should be in format
"""
description blha blha
@param1
@param2
@returns 
"""

File: BST/Node.py [hunk 10]
```diff
@@ -0,0 +1,84 @@
...
+    def actual_search(self, treasure):
```
Comment (ofekba) [inline BST/Node.py]: add to all functions input types + output type

File: BST/Node.py [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+        # Returning if node is found
+        if float(treasure) == self.treasure:
+            return ["", self]
+        # Progressing down the tree or returning if the end has been reached
+        if float(treasure) < self.treasure:
+            if self.left is None:
+                return ["left", self]
```
Comment (ofekba) [inline BST/Node.py]: use constants instead of text

File: BST/Node.py (hunk 10 above)
Comment (ofekba) [inline BST/Node.py]: הקוד פה לא ממש קריא תנסה לכתוב מחדש מסודר יותר

File: BST/Node.py [hunk 12]
```diff
@@ -0,0 +1,84 @@
...
+            res = self.left.actual_search(treasure)
+            res.append(self)
+            return res
+        if self.right is None:
+            return ["right", self]
+        res = self.right.actual_search(treasure)
+        res.append(self)
+        return res
+    # Returning the path to the successor in the right subtree of a node
```
Comment (ofekba) [inline BST/Node.py]: 2 empty lines should be as separation between different functions

File: BST/Node.py [hunk 13]
```diff
@@ -0,0 +1,84 @@
...
+    def find_successor(self):
+        # Going left until finding a node without a left child
+        if self.left is None:
+            return [self]
+        res = self.left.find_successor()
+        res.append(self)
+        return res
+    # Passing throughout the tree and printing the nodes based on the input order
+    def bst_pass(self, order, node_list):
```
Comment (ofekba) [inline BST/Node.py:74]: separate to functions.

File: BST/Node.py [hunk 14]
```diff
@@ -0,0 +1,84 @@
...
+        if self.right is not None:
+            node_list = self.right.bst_pass(order, node_list)
+        if order == "post-order":
+            node_list.append(self.treasure)
+            print("current treasure: " + str(self.treasure))
+        return node_list
+
+    # Passing throughout the tree, checking if the nodes are ordered correctly and returning the tree's depth
+    def validate(self, depth):
+        if (self.left is not None and self.left.treasure > self.treasure)\
```
Comment (ofekba) [inline BST/Node.py]: write shortly is not None

File: BST/Tree.py [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+    # Check if the tree is empty, else search the node in the tree
+    def actual_search(self, treasure):
+        if self.root is None:
+            return ["", None]
+        return self.root.actual_search(treasure)
+    # Search the node, extract it from the path and print a relevant message
+    def search(self, treasure):
+        res = self.actual_search(treasure)
+        if res[0] != "" or res[1] is None:
+            print("treasure " + str(treasure) + " is not in the bst")
```
Comment (ofekba) [inline BST/Tree.py]: use logger for logging

File: BST/Tree.py [hunk 16]
```diff
@@ -0,0 +1,225 @@
...
+        db_node = table.find({"treasure": value})
+        for node in db_node:
+            if node["left"] != "":
+                new_node.left = self.rebuild_node(str(node["left"]), table)
+            if node["right"] != "":
+                new_node.right = self.rebuild_node(str(node["right"]), table)
+        return new_node
+
+    # Inserting the whole tree from MongoDB
+    def rebuild(self, table):
```
Comment (ofekba) [inline BST/Tree.py]: change all function names to be more indictive

File: BST/Tree.py [hunk 17]
```diff
@@ -0,0 +1,225 @@
...
+        new_tree = Tree()
+        db_tree = table.find({"class": "tree"})
+        if table.count_documents({"class": "tree"}) == 0:
+            return new_tree
+        for tree in db_tree:
+            if tree["root"] == "":
```
Comment (ofekba) [inline BST/Tree.py]: if root key not exists?

File: BST/Tree.py [hunk 18]
```diff
@@ -0,0 +1,225 @@
...
+                "class": "vis",
+                "content": vis
+            })
+        query = {"class": "vis"}
+        update = {"$set": {"content": vis}}
+        table.update_one(query, update)
+
+    # Getting the visualization format from MongoDB
+    def get_vis(self, table):
+        db_res = table.find({"class": "vis"})
```
Comment (ofekba) [inline BST/Tree.py]: if key not exists what will happend?

File: BST/Node.py [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+    # Initializing empty node
+    def __init__(self, value):
+        self.left = None
+        self.right = None
+        self.treasure = value
+    # Function returns the path to a node if exist else return the path to the place in which the node will be added
+    def actual_search(self, treasure):
+        # Returning if node is found
+        if float(treasure) == self.treasure:
+            return ["", self]
```
Comment (MayaLandman) [inline BST/Node.py]: No need to return self, you call the function by the original self

File: BST/BST_task.py (hunk 1 above)
Comment (Alon6) [reply BST/BST_task.py]: MongoDB creates the table and the db if they did not exist

File: BST/Node.py (hunk 19 above)
Comment (Alon6) [reply BST/Node.py]: You need to return the self of the node we are searching, the original self is the root

File: BST/Tree.py (hunk 17 above)
Comment (Alon6) [reply BST/Tree.py]: Documents of class tree are created with a "root" property

File: BST/BST_task.py [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+    return "Error - Delete query has failed", HTTPStatus.BAD_REQUEST
+
+
+@app.route('/insert_treasure', methods=['POST'])
+def insert_treasure():
+    val_json = request.get_json()
+    if "value" not in val_json:
+        return "Error, input does not contain treasure", HTTPStatus.BAD_REQUEST
+    val = val_json["value"]
+    res = bst.insert(Node(val), table)
```
Comment (MayaLandman) [inline BST/BST_task.py]: Check for duplicate inserts before inserting treasure

File: BST/BST_task.py [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+    if "value" not in val_json:
+        return "Error, input does not contain treasure", HTTPStatus.BAD_REQUEST
+    val = val_json["value"]
+    res = bst.delete(val, table)
+    if res[0]:
+        logger.info("Treasure " + str(val) + " was deleted successfully")
+        return "Success", HTTPStatus.OK
+    else:
+        logger.error(res[1])
+        return "Error - " + res[1], HTTPStatus.BAD_REQUEST
```
Comment (MayaLandman) [inline BST/BST_task.py:71]: You may want to return NOT_FOUND to a missing delete

File: BST/Node.py [hunk 22]
```diff
@@ -0,0 +1,81 @@
...
+class Node:
+    # Initializing empty node
+    def __init__(self, value):
+        self.left = None
+        self.right = None
+        self.treasure = value
+    # Function returns the path to a node if exist else return the path to the place in which the node will be added
+    def actual_search(self, treasure):
+        # Returning if node is found
+        if float(treasure) == self.treasure:
```
Comment (MayaLandman) [inline BST/Node.py:38]: Validate input data before casting to float to avoid errors

File: BST/Tree.py [hunk 23]
```diff
@@ -0,0 +1,246 @@
...
+        self.root = None
+    # Check if the tree is empty, else search the node in the tree
+    def actual_search(self, treasure):
+        if not self.root:
+            return ["", None]
+        return self.root.actual_search(treasure)
+    # Search the node, extract it from the path and print a relevant message
+    def search(self, treasure):
+        res = self.actual_search(treasure)
+        if res[0] != "" or not res[1]:
```
Comment (MayaLandman) [inline BST/Tree.py]: if res[0] != "" can be if res[0]

File: BST/Tree.py [hunk 24]
```diff
@@ -0,0 +1,246 @@
...
+                flag = flag and self.update_root_db(child, table)
+            else:
+                flag = flag and self.swap(node, child, parent, table)
+        # If the node has two children then swap the node with its successor (and if the successor has a child then
+        # replace the successor with its child
+        else:
+            suc_res = node.right.find_successor()
+            suc = suc_res[0]
+            suc.left = node.left
+            flag = flag and self.update_node_db(suc, node.left, "left", table)
```
Comment (MayaLandman) [inline BST/Tree.py]: consider defining constants for "left" and "right"

File: BST/Tree.py [hunk 25]
```diff
@@ -0,0 +1,246 @@
...
+            update = {"$set": {dir: str(child.treasure)}}
+            table.update_one(query, update)
+            return True
+        except Exception as e:
+            return False
+
+    # Inserting a node in MongoDB
+    def insert_db(self, new_node, father_node, table, direction):
+        try:
+            db_tree = table.find({"class": "tree"})
```
Comment (MayaLandman) [inline BST/Tree.py]: Could check for no docs and do the insert, then you can use find only once

File: BST/BST_task.py (hunk 20 above)
Comment (Alon6) [reply BST/BST_task.py]: This is already implemented in the insert function

File: BST/BST_task.py [hunk 26]
```diff
@@ -0,0 +1,165 @@
+from http import HTTPStatus
+from Node import Node
+from flask import Flask, request
+from pymongo import MongoClient
+from Tree import Tree
+import logging
+
+# Initializing server and BST variables
+app = Flask(__name__)
+client = MongoClient("localhost", 27017)
```
Comment (NadavGlazer) [inline BST/BST_task.py]: move connection details to env file

Comment (NadavGlazer) [inline BST/BST_task.py]: Move the mongo connection to anther file

File: BST/BST_task.py [hunk 27]
```diff
@@ -0,0 +1,165 @@
...
+db = client["bst_db"]
+table = db["bst_table12"]
+bst = Tree()
+
+# Create and configure logger
+logging.basicConfig(filename="log.log",
```
Comment (NadavGlazer) [inline BST/BST_task.py]: move the logging to anther file, call only the initialization here

File: BST/BST_task.py (hunk 26 above)
Comment (NadavGlazer) [inline BST/BST_task.py]: create a class that communicates with mongo

File: BST/BST_task.py [hunk 28]
```diff
@@ -0,0 +1,165 @@
...
+        logger.info("Deleted table content")
+        return "Success", HTTPStatus.OK
+    logger.error("Table content deletion failed")
+    return "Error - Delete query has failed", HTTPStatus.BAD_REQUEST
+
+
+@app.route('/insert_treasure', methods=['POST'])
+def insert_treasure():
+    val_json = request.get_json()
+    if "value" not in val_json or not is_float(str(val_json["value"])):
```
Comment (NadavGlazer) [inline BST/BST_task.py]: simplify the condition- make it more understandable

File: BST/BST_task.py [hunk 29]
```diff
@@ -0,0 +1,165 @@
...
+        logger.error("Error, input does not contain treasure")
+        return "Error, input does not contain treasure", HTTPStatus.BAD_REQUEST
+    val = val_json["value"]
```
Comment (NadavGlazer) [inline BST/BST_task.py]: you have already got the value in the condition above

File: BST/BST_task.py [hunk 30]
```diff
@@ -0,0 +1,165 @@
...
+    res = bst.insert(Node(val), table)
+    if res[0]:
+        logger.info("Treasure " + str(val) + " was inserted successfully")
```
Comment (NadavGlazer) [inline BST/BST_task.py]: write the log using f string

File: BST/BST_task.py [hunk 31]
```diff
@@ -0,0 +1,165 @@
...
+
+# Create and configure logger
+logging.basicConfig(filename="log.log",
+                    format='%(asctime)s %(message)s',
+                    filemode='w')
+
+# Creating an object
+logger = logging.getLogger()
+logger.setLevel(0)
+def is_float(string):
```
Comment (NadavGlazer) [inline BST/BST_task.py]: move to a utils file

File: BST/BST_task.py [hunk 32]
```diff
@@ -0,0 +1,165 @@
...
+
+@app.route('/get_treasures')
+def get_treasures():
+    val = bst.bst_pass("in-order")
+    logger.info("The pass was a success")
+    return {"treasures": val}, HTTPStatus.OK
+
+
+@app.route('/delete_treasure', methods=['DELETE'])
+def delete_treasure():
```
Comment (NadavGlazer) [inline BST/BST_task.py:54]: documentation - what does the function receive and what it does
